### PR TITLE
Refactor Penthesilea and Dorothea

### PR DESCRIPTION
**Pull request by jacg**
This is an issue which will eventually turn into a PR. Given that GitHub has removed the ability of upgrading issues to PRs, I have started the issue as a PR, in order to save us the need of having both an issue and a PR covering the same topic.

The abstract methods `file_loop`, `event_loop`, `write_parameters` and `get_writers` are not consistent: Two of them raise very specific (and distinct) IC exceptions whose meaning is `NotImplementedError`; two of them `pass`.

This should be made consistent in one of two ways:

1. Remove the IC-specific exceptions, and make all 4 methods raise `NotImplementedError`.
2. Create two more IC-specific exceptions and make all 4 of them subclasses of `NotImplementedError` (as well as as `ICException`), and make each method raise its specific exception.

Once we agree on which option is preferred, anyone should feel free to effect the changes: I am unlikely to be around to do it myself.

### CODE REVIEW

**Comment by gonzaponte** [comment]
Are specific exceptions giving any extra information?

**Comment by jacg** [comment]
Psah, they make the problem and its solution and bit more obvious and in-your-face. I'm not convinced that it's better than simply raising `NotImplementedError` with clear message inside it.

As a fun challenge / bonus, we could include the name of the class in which the method should be overridden, in the message!

**Comment by jjgomezcadenas** [comment]
what is the status of this?

**Comment by jacg** [comment]
As the introductory comment and subsequent discussion suggest, no work has been done on this yet. It is an issue waiting to be turned into a PR when someone finds the time to do something about it. It's trivially easy to do, but I have about 46 higher priority IC things on my hands at the moment, not to mention non-IC things.

**Comment by jacg** [comment]
This one is ready too, now.

It's trivial.

**Comment by jacg** [comment]
Hmm, no, it's not ready!

Apparently Dotty and Izzy do not implement `wirte_parameters`. Is that supposed to be allowed?

**Comment by jjgomezcadenas** [comment]
Patty neither. It turns out that only DIOMIRA implements write parameters. This needs to be fixed (should not be in the interface is a very local thing of DIOMIRA). Easy to fix.

**Comment by jacg** [comment]
> Patty neither.

I didn't see any Patty tests failing because of it. Which would suggest that Patty doesn't *call* `write_parameters`?

**Comment by jjgomezcadenas** [comment]
As I said, only DIMORA calls write_parameters. The function should disappear from the interface, I think.

**Comment by jacg** [comment]
> This needs to be fixed (should not be in the interface is a very local thing of DIOMIRA). Easy to fix.

As you seem to know what the expectations and purposes are, please do it. Or delegate to someone who does. I don't think I have anything left to contribute to this PR.

**Comment by jjgomezcadenas** [comment]
I have some time to work on this now. I can improve a bit the messy event_loop, at least unifiying the code of Patty and Dotty. Shall I carry on that work on this PR, @jacg ?

**Comment by jacg** [comment]
I think it belongs in a separate PR. Unless you feel that completing this PR is effectively impossible without unifying Patty and Dotty. In which case the title of this PR should be changed, because the other works seems much more significant than the trivial point that this PR originally addresses.

**Comment by jjgomezcadenas** [comment]
Well, I pulled your branch and I am almost done with a major refactoring of P&D, which have lost about 50 pounds each. I will submit a separate PR, but I am already incorporating your commits to my PR (there were necessary, any way).

**Comment by jacg** [comment]
Don't bother making a new PR, just push your work to this one. I have already changed the title. If you don't like the title, click the `edit` button waaaay to the right of the title.

**Comment by jjgomezcadenas** [comment]
great! Thanks Master

**Comment by jjgomezcadenas** [comment]
Initial review of this is already possible. How busy are you @jacg? Depending on the schedule of master, @abotas can review it (take a look anyway, will you?)

**Comment by jacg** [comment]
> How busy are you @jacg?

Essentially unavailable. Might find some time on Monday. Can't promise it though. (I'm having a cursory glance right now, but it will be very superficial for now.)

**Comment by jjgomezcadenas** [comment]
>Essentially unavailable. Might find some time on Monday. Can't promise it though. (I'm having a cursory glance right now, but it will be very superficial for now.)

OK, unless you say otherwise, I will pass the torch to @abotas after your initial revision.

**Comment by jacg** [comment]
Don't delay @abotas, waiting for me. Don't expect any input from me! I'm just glancing over it to have the vaguest of ideas of what you've touched, in case I can join in later.

**Comment by jjgomezcadenas** [comment]
Agreed then. @abotas, please take a look to the PR! Thanks!

**Comment by jacg** [comment]
Rhetorical question: Are you expecting `test_command_line_run[penthesilea]` to fail?

**Comment by jjgomezcadenas** [comment]
Certainly not (it is not failing in my local and apparently Travis not failing either)

**Comment by jjgomezcadenas** [comment]
... and it should not fail. Penthesilea and Dorothea are now doing almost nothing, all common funcionality moved to a common event_loop, reducing  noise and hopefully spureous errors.

**Comment by jacg** [comment]
Hmm. I have re-downloaded the database, cleaned, re-compiled, and I consistently get a segfault in `test_command_line_run[penthesilea]`. Running Penthesilea manually on the command line also gives a segfault. (And that's all I have time for, for now.)

**Comment by abotas** [comment]
All tests pass on my machine and on Travis.

**Review by abotas — APPROVED** [review]
This PR simplifies and improves a few of our cities and base cities. Replacing the IC exceptions with `NotImplementedError` is straight forward and concise. Dorothea and Penthesilea have a lot in common and their event loops showed it. This PR squashes redundancies into a more versatile `PCity`. Penthesilea and Dorothea now need only worry themselves with setting writers and passing them their specific `dst_event` which are then used by `PCity`'s event loop. Looks much better to me -- Approved.

**Comment by jacg** [comment]
Reinstalling the IC3.6 environment from scratch made the local segfault go away. All tests pass here, now.

File: invisible_cities/cities/penthesilea.py [hunk 1]
```diff
@@ -3,91 +3,25 @@
...
-                                      timestamps = evt_time)
-            evt = self.create_hits_event(pmapVectors)
-            write_hits(evt)
 
     def get_writers(self, h5out):
         """Get the writers needed by dorothea"""
         return  hits_writer(h5out)
+
+    def create_dst_event(self, pmapVectors):
+        """Get the writers needed by dorothea"""
```
Comment (jacg) [inline invisible_cities/cities/penthesilea.py:23]: The docstring of `Penthesilea.create_dst_event` is "Get the writers needed by dorothea".

1. I really dislike these vacuous docstrings: if they don't say anything that isn't blatantly obvious from the name of the function, what's the point of having them?
2. If on top of being pointless, they are also *wrong* ...

Could we please drop the habit of writing utterly pointless docstrings?

I was hoping to clear out this sort of rubbish *before* this PR was accepted.

**Pull request merged**